src/scattering/particle.h

```diff
@@ -0,0 +1,617 @@
+        properties_(ParticleProperties{"", "", "", mass, d_eq, d_max, 0.0}) {}
+
+  Particle(const Particle&) = default;
+  Particle& operator=(const Particle&) = default;
+
+  Particle copy() const { return Particle(properties_, data_.copy()); }
+
+  //
+  // Particle meta data.
+  //
+
+  /// The name of the particle, if available. Empty string otherwise.
+  std::string get_name() const { return properties_.name; }
+  /// The source of the particle data, if available. Empty string otherwise.
+  std::string get_source() const { return properties_.source; }
+  /// The refractive index of the particle data, if available. Empty string otherwise.
+  std::string get_refractive_index() const { return properties_.refractive_index; }
+  ParticleType get_particle_type() const { return data_.get_particle_type(); }
+  DataFormat get_data_format() const { return data_.get_data_format(); }
+  /// The number of stokes dimensions of the scattering data.
+  Index get_stokes_dim() const { return data_.get_stokes_dim(); }
+
+  //
+  // Particle size and mass.
+  //
+
+  /// The particle mass.
+  double get_mass() const { return properties_.mass; }
+  /// The particle maximum diameter.
+  double get_d_max() const { return properties_.d_max; }
+  /// The particle volume-equivalent diameter.
+  double get_d_eq() const { return properties_.d_eq; }
+  /// The aerodynamic cross-sectional area.
+  double get_d_aero() const { return properties_.d_aero; }
+
+  //
+  // Scattering data.
+  //
+
+  /// The frequency grid over which the data is defined.
+  const math::Vector<double>& get_f_grid() const { return data_.get_f_grid(); }
+  /// The temperature grid over which the data is defined.
+  const math::Vector<double>& get_t_grid() const { return data_.get_t_grid(); }
+  /// Longitudinal component of the incoming angle.
+  math::Vector<double> get_lon_inc() const { return data_.get_lon_inc(); }
+  /// Latitudinal component of the incoming angle.
+  math::Vector<double> get_lat_inc() const { return data_.get_lat_inc(); }
+  /// Longitudinal component of the scattering (outgoing) angle.
+  math::Vector<double> get_lon_scat() const { return data_.get_lon_scat(); }
+  /// Latitudinal component of the scattering (outgoing) angle.
+  math::Vector<double> get_lat_scat() const { return data_.get_lat_scat(); }
+
+  //////////////////////////////////////////////////////////////////////////////
+  // Manipulation of scattering data
+  //////////////////////////////////////////////////////////////////////////////
+
+  // pxx :: hide
+  /** Frequency interpolation.
+   *
+   * Linear interpolation of scattering data along frequency dimension.
+   *
+   * @param f_grid The frequency grid to interpolate the data to.
+   * @return New particle with the scattering data interpolated to the
+   * given frequencies.
+   */
+  Particle interpolate_frequency(
+      std::shared_ptr<math::Vector<double>> f_grid) const {
+    return Particle(properties_, data_.interpolate_frequency(f_grid));
```

Comment:
If I understand the documentation of ``std::enable_shared_from_this`` correctly, I wouldn't help with the type signatures here. Moreover, isn't creating a shared pointer out of a reference just the misuse good example from the [docs](https://en.cppreference.com/w/cpp/memory/enable_shared_from_this)?

The motivation for using shared pointers here is actually two-fold:

1. This should save some memory since particles will finally be brought to the same grids, so it feels natural that particles on the same grid should share those vectors. I admit, however, that this is likely a case of premature optimization.
2. More importantly, using sharing the ownership makes it easier to discover if all data has been brought to the same grid.

Reply:
It is perhaps the misuse case.

I think I formulated myself poorly.  I understand that the `*this` class is never value constant, so the shared pointer may modify some of the underlying values.  Is it?

Because I feel like I want to interpolate to a grid or value, not to a pointer.  Sending in a shared pointer indicates I am not only interpolating but also doing something else.  So either the name of the function is wrong (perhaps it should be `interpolate_frequency_and_update_data`), or the function can take a constant reference to new frequency grid.

With the current name, and if you need to update the data but this will be done in such a way that the `Particle` that is returned is practically value-constant, using `enabled_shared_from_this` allows you to hide that detail while still keeping the name.  This was mostly to work around the name-understanding problem I had here

Reply:
`enable_shared_from_this` would never be undefined behavior.  It always throws quite reliably.  Anyways, this is besides the point, it's fine you don't want to use it.

As I understand how this part of the code is intended to work without fault: you have some function that creates a bunch of `Particle`.  A single `Particle` needs to live on a frequency grid.  You might want many more `Particle` than frequency grids, so you let them share frequency grids.  This shared frequency grids are created as `const math::Vector<double>`, since you don't want it to change as you have already done the interpolation.  The original `const math::Vector<double>` pointers may then all live on after the function that created the `Particle`.

Now my understanding fails me: How can a shared pointer to a `math::Vector<double>` be guaranteed const valued?  If it is not const, then it seems you are at a risk of changing the grid values even though you've already interpolated the values.  This matters to me because there are two frequency grids in ARTS.  One sensor-frequency grid `f_grid` that users set, and then an internal Doppler shifted version that changes per atmospheric level due to wind.  It is the latter you likely want to use in these interpolations.

Reply:
OK, so you just want me to the pointee vector ``const``?

Reply:
If that is possible.  It allows all of my problems of understanding to go away :)